### ISSUE TITLE
HL-2230 graphics are cut off

### ISSUE THREAD

**Issue by jclouter** [pdewacht/brlaser#95]
This printer seems to work fairly well when printing text, however pages featuring large graphics are typically cut off towards the top. It's not on the list of supported printers, however using the 'Brother HL-1110 series, using brlaser v6' option seems to allow partial functionality as long as there's no large graphics. If the pages are scaled down to about 70 percent they sometimes will print. Would be great to have full HL-2230 support!

**Comment by jclouter** [comment]
Sorry should have mentioned:

Linux raspberrypi 5.4.72-v7+ #1356 SMP
CUPS 2.2.10
brlaser v6

**Comment by alexivkin** [comment]
Same exact issue with rpi on a similar printer. And same exact workaround by scaling images down.
Are you using a compiled driver from this repo or one from the apt repo?

**Comment by jclouter** [comment]
> Same exact issue with rpi on a similar printer. And same exact workaround by scaling images down.
> Are you using a compiled driver from this repo or one from the apt repo?

Compiled it from this repo

**Comment by jclouter** [comment]
I've also tried editing the PPD file for the HL-1110 to use the page area / printable area that is included in Brother's official HL-2230 PPD file. That Brother file is incompatible with cups on raspberry pi, but my edited file works when I use it. The same problem persists however, anything with large graphics is cut off or blank. Insufficient memory somewhere perhaps?

**Comment by alexivkin** [comment]
There are no issues like this when printing on x86 with the native i386 Brother drivers. It seems native drivers stream data to work on these printers, unlike brlaser which tries to stuff everything in the little memory this device has. 

Good news - There are armv7l compiled drivers from Brother which may work for y.
As for brlaser, @pdewacht perhaps there is something to be gleaned from the [reverse engineered rawtobr3](https://github.com/k1-801/rawtobr3)?

**Comment by jclouter** [comment]
> There are no issues like this when printing on x86 with the native i386 Brother drivers. It seems native drivers stream data to work on these printers, unlike brlaser which tries to stuff everything in the little memory this device has.
> 
> Good news - There are armv7l compiled drivers from Brother which may work for y.
> As for brlaser, @pdewacht perhaps there is something to be gleaned from the [reverse engineered rawtobr3](https://github.com/k1-801/rawtobr3)?

Yes agreed, I'm currently using an old Ubuntu machine as my cups print server and the printer works perfectly with that setup. A raspberry Pi would be preferable so I don't have to have that noisy old machine on 24/7 :-) The reason for the print server is so that my kids can print their schoolwork assignments from their Chromebooks which don't support the HL-2230. 

I wasn't able to find the ARM compiled drivers from Brother you spoke of - any chance you can direct me to them? Thanks!

**Comment by alexivkin** [comment]
Try the steps below. It's written for HL2270DW, so correct the steps for HL-1110 
### Repackage LPR drivers

Extract the brother ARM drivers from a generic armhf archive they provide.

	wget http://download.brother.com/welcome/dlf103361/brgenprintml2pdrv-4.0.0-1.armhf.deb
	dpkg -x brgenprintml2pdrv-4.0.0-1.armhf.deb brgenprintml2pdrv-4.0.0-1.armhf.extracted
	dpkg-deb -e brgenprintml2pdrv-4.0.0-1.armhf.deb brgenprintml2pdrv-4.0.0-1.armhf.extracted/DEBIAN

Get the original i386 LPR driver, unpack the files and the control information

	wget https://download.brother.com/welcome/dlf005893/hl2270dwlpr-2.1.0-1.i386.deb
	dpkg -x hl2270dwlpr-2.1.0-1.i386.deb hl2270dwlpr-2.1.0-1.armhf.extracted
	dpkg-deb -e hl2270dwlpr-2.1.0-1.i386.deb hl2270dwlpr-2.1.0-1.armhf.extracted/DEBIAN

Now copy the ARM code into the unpacked folders

	cp brgenprintml2pdrv-4.0.0-1.armhf.extracted/opt/brother/Printers/BrGenPrintML2/lpd/armv7l/brprintconflsr3 hl2270dwlpr-2.1.0-1.armhf.extracted/usr/local/Brother/Printer/HL2270DW/inf
	cp brgenprintml2pdrv-4.0.0-1.armhf.extracted/opt/brother/Printers/BrGenPrintML2/lpd/armv7l/rawtobr3 hl2270dwlpr-2.1.0-1.armhf.extracted/usr/local/Brother/Printer/HL2270DW/inf

And repackage it

	cd hl2270dwlpr-2.1.0-1.armhf.extracted
	find . -type f ! -regex '.*.hg.*' ! -regex '.*?debian-binary.*' ! -regex '.*?DEBIAN.*' -printf '%P ' | xargs md5sum > DEBIAN/md5sums
	chmod 755 hl2270dwlpr-2.1.0-1.armhf.extracted/DEBIAN/p* hl2270dwlpr-2.1.0-1.armhf.extracted/usr/local/Brother/Printer/HL2270DW/inf/* hl2270dwlpr-2.1.0-1.armhf.extracted/usr/local/Brother/Printer/HL2270DW/lpd/*
	cd ..
	dpkg-deb -b hl2270dwlpr-2.1.0-1.armhf.extracted hl2270dwlpr-2.1.0-1.armhf.deb

### Repackage CUPS wrapper

Grab and extract the `brcupsconfig4` sources

	wget https://download.brother.com/welcome/dlf006733/brhl2270dwcups_src-2.0.4-2.tar.gz
	tar zxvf brhl2270dwcups_src-2.0.4-2.tar.gz
	cd brhl2270dwcups_src-2.0.4-2

On arm64 you need to cross-compile as below. On arm32 use `gcc` instead.

 	sudo apt install gcc-9-arm-linux-gnueabihf
	arm-linux-gnueabihf-gcc-9 brcupsconfig3/brcupsconfig.c -o brcupsconfig4

Grab the original i386 CUPS wrapper and unpack it

	wget https://download.brother.com/welcome/dlf005893/cupswrapperHL2270DW-2.0.4-2.i386.deb
	dpkg -x cupswrapperHL2270DW-2.0.4-2.i386.deb cupswrapperHL2270DW-2.0.4-2.armhf.extracted
	dpkg-deb -e cupswrapperHL2270DW-2.0.4-2.i386.deb cupswrapperHL2270DW-2.0.4-2.armhf.extracted/DEBIAN

Copy the compiled code into the unpacked folder

	cp brhl2270dwcups_src-2.0.4-2/brcupsconfig4 cupswrapperHL2270DW-2.0.4-2.armhf.extracted/usr/local/Brother/Printer/HL2270DW/cupswrapper

Repack it

	cd cupswrapperHL2270DW-2.0.4-2.armhf.extracted
	find . -type f ! -regex '.*.hg.*' ! -regex '.*?debian-binary.*' ! -regex '.*?DEBIAN.*' -printf '%P ' | xargs md5sum > DEBIAN/md5sums
        chmod 755 cupswrapperHL2270DW-2.0.4-2.armhf.extracted/DEBIAN/p* cupswrapperHL2270DW-2.0.4-2.armhf.extracted/usr/local/Brother/Printer/HL2270DW/cupswrapper/*
	dpkg-deb -b cupswrapperHL2270DW-2.0.4-2.armhf.extracted cupswrapperHL2270DW-2.0.4-2.armhf.deb
